### ISSUE TITLE
Stabilize existing neon instrinsics

### ISSUE THREAD

**Issue by jrmuizel** [rust-lang/stdarch#1125]
Having to use nightly for NEON is pain. Is there any reason we can stabilize the intrinsics that we already have implemented? If not, what's needed to do this?

**Comment by Lokathor** [comment]
opened a zulip topic to get this a little more attention: https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/When.20can.20we.20do.20a.20partial.20stabilization.20of.20Neon.20intrinsics.3F

**Comment by Amanieu** [comment]
One thing that would make me more comfortable with partial stabilization is a test that checks that the stdarch intrinsics behave identically to the C versions with random inputs.

**Comment by jrmuizel** [comment]
Is there some infrastructure for doing that on x86 that can be reused?

**Comment by Amanieu** [comment]
No. Ideally we would want a solution that can auto-generate tests for both ARM and x86.

**Comment by SparrowLii** [comment]
> One thing that would make me more comfortable with partial stabilization is a test that checks that the stdarch intrinsics behave identically to the C versions with random inputs.

How do we get the behavior of these instructions in Clang? Can it be a way to get them from llvm's documentation?

**Comment by JamieCunliffe** [comment]
As mentioned on zulip, we had been looking at alternative ways to generate the intrinsics, one of the things we did could be repurposed to compare c and rust intrinsics. What we did was basically generate C programs and compile them to LLVM IR so it wouldn't be much more to actually run them to get the results.

I can change the input source to be the spreadsheet that is currently tracking intrinsics, that should make it fairly generic.

My initial thoughts were this would be a binary that was run that would generate some C and rust that calls the intrinsic with the same random inputs and then print the results of the intrinsic call, we can then diff the two programs outputs and see if anything differs. If that's the case I should hopefully be able to share something soon.

**Comment by Amanieu** [comment]
> My initial thoughts were this would be a binary that was run that would generate some C and rust that calls the intrinsic with the same random inputs and then print the results of the intrinsic call, we can then diff the two programs outputs and see if anything differs. If that's the case I should hopefully be able to share something soon.

Yes that's exactly what I had in mind.

**Comment by bartelsielski** [comment]
https://github.com/rust-lang/stdarch/pull/1266 stabilized the neon intrinsics for aarch64. Are there plans to also stabilize neon on 32-bit arm?

**Comment by Amanieu** [comment]
There are many intrinsics that are still buggy or missing on ARM: https://github.com/rust-lang/stdarch/blob/master/crates/intrinsic-test/missing_arm.txt

**Comment by RalfJung** [comment]
> [#1266](https://github.com/rust-lang/stdarch/pull/1266) stabilized the neon intrinsics for aarch64. Are there plans to also stabilize neon on 32-bit arm?

Unfortunately those intrinsics are broken on the LLVM side: https://github.com/rust-lang/rust/issues/129880.
